src/services/registrations.ts

```diff
@@ -8,21 +8,80 @@ import {
   upsertQuestionOptionFromRegistrationData,
 } from './registrationData';
 
-export async function saveRegistration(
-  dbPool: PostgresJsDatabase<typeof db>,
-  data: z.infer<typeof insertRegistrationSchema>,
-  userId: string,
-) {
-  if (data.groupId) {
+export async function validateCreateRegistrationPermissions({
+  dbPool,
+  userId,
+  groupId,
+  eventId,
+}: {
+  dbPool: PostgresJsDatabase<typeof db>;
+  userId: string;
+  eventId: string;
+  groupId?: string | null;
+}) {
+  if (groupId) {
+    const userGroup = dbPool.query.usersToGroups.findFirst({
+      where: and(eq(db.usersToGroups.userId, userId), eq(db.usersToGroups.groupId, groupId)),
+    });
+
+    if (!userGroup) {
+      return false;
+    }
+
+    // limit one registration per group per event
+    const existingRegistration = await dbPool.query.registrations.findFirst({
+      where: and(eq(db.registrations.eventId, eventId), eq(db.registrations.groupId, groupId)),
+    });
+
+    if (existingRegistration) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+export async function validateUpdateRegistrationPermissions({
+  dbPool,
+  registrationId,
+  userId,
+  groupId,
+}: {
+  dbPool: PostgresJsDatabase<typeof db>;
+  userId: string;
+  registrationId: string;
+  groupId?: string | null;
+}) {
+  const existingRegistration = await dbPool.query.registrations.findFirst({
+    where: and(eq(db.registrations.userId, userId), eq(db.registrations.id, registrationId)),
+  });
+
+  if (!existingRegistration) {
+    return false;
+  }
+
+  if (existingRegistration.userId !== userId) {
+    return false;
+  }
```

Comment:
Add comment like: "// Only creator of the registration can update the registration" as this line is the key line that ensures this.